src/Tools/LinkabilityChecker/LinkabilityChecker.csproj

```diff
@@ -22,6 +22,12 @@
       <!-- Link all assemblies -->
       <ILLinkArgs>$(ILLinkArgs) --action link</ILLinkArgs>
       <ILLinkArgs Condition="'$(GenerateLinkerWarningSuppressions)' == 'true'">$(ILLinkArgs) --generate-warning-suppressions xml</ILLinkArgs>
+      <!-- suppress warnings with the following codes:
+           IL2026: Calling A which has B can break functionality when trimming application code. The target method might be removed.
```

Comment:
I guess the trimmer produces warnings if you have `RequiresUnreferencedCode` in your code which we do. The runtime also appears to suppress a bunch of these warnings, maybe these are primarily for applications and not for libraries?

Reply:
It should not. It should only produce a warning if you have code which is not annotated which calls into something which has RUC on it.

This is VERY likely hiding a trimming hole somewhere - please don't do this!

Runtime does suppress some of these warnings, but it does so selectively at the point where they occur and we spent a LOT of time justifying the suppression.

In general my advice would be - treat any suppression as a bug. It can be a bug in the linker (like the various workaround for lambdas and so on) or bug in the annotated code, but it's almost always a bug. The cases where it's not are rare and require deep understanding of the area and the trimmer behavior (which we typically don't want to take a dependency on anyway).

Reply:
> See also: https://github.com/dotnet/runtime/blob/7671b5084a41944d7a0c1fb9674e354183ec722d/eng/illink.targets#L238-L255

Note that we do different types of trimming in the runtime repo, so it isn't useful to compare.

In the case that you pointed to, we only run a single assembly through the trimmer and trim out any dead code. This isn't our "trim analysis" step - that step is separate and runs on the whole shared framework together. In that step we don't globally suppress any warnings. That code is here:

https://github.com/dotnet/runtime/blob/7671b5084a41944d7a0c1fb9674e354183ec722d/src/libraries/sfx.proj#L47-L63

Reply:
What Eric says - that run of trimmer has basically analysis disabled because we don't want any of the warnings from it.

Reply:
If I remove these warnings, I get a bunch of warnings I wasn't sure how to address. Here they are:

```C#
Severity	Code	Description	Project	File	Line	Suppression State
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'Microsoft.Extensions.Internal.PropertyHelper.<GetProperties>g__AddInterestingProperties|41_0(List<PropertyHelper>,Type)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'Microsoft.Extensions.Internal.PropertyHelper.GetProperties(Type)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'Microsoft.Extensions.Internal.PropertyHelper.GetVisibleProperties(Type)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.CreateDelegate(Type,Type,String)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.CreateDelegate(Type,Type,String,Boolean)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.CreateDelegate(Type,Type,String,Boolean,Boolean)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.Delegate(Type,String)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.MulticastDelegate.MulticastDelegate(Type,String)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
```

Reply:
I also get warnings for things which I thought have the right annotations, but the trimmer still warns:

```C#
Severity	Code	Description	Project	File	Line	Suppression State
Warning	IL2026	Microsoft.AspNetCore.Http.Abstractions: Using member 'Microsoft.AspNetCore.Routing.RouteValueDictionary.PropertyStorage.PropertyStorage(Object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This API is not trim safe.
Severity	Code	Description	Project	File	Line	Suppression State
Warning	IL2026	Microsoft.AspNetCore.Http.Abstractions: Using member 'Microsoft.AspNetCore.Routing.RouteValueDictionary.RouteValueDictionary(Object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.
```

Both of these methods have `RequiresUnreferencedCode`. Similarly a bunch of things in `PropertyHelpers` which are annotated with `RequiresUnreferencedCode` appear in the warning list.

Reply:
I looked into this a bit more.
The IL2111 on PropertyHelper is caused by the `MetadataUpdateHandler(typeof(PropertyHelper))`. The attribute has annotation `All` on the type - so it will force include everything on the class - and if there's something which has annotation it will warn about it. I don't know why it needs `All` - based on the description in the docs, it should be enough to have "all methods" (not that it would help here).
I can't think of a good way to fix this . You could suppress the warning on the entire type (PropertyHelper) with an explanation that it's because of the MetadataUpdateHandler and that it will only ever use the `ClearCache` method which doesn't have any annotation.
Alternatively, you could create a static subclass like `PropertyHelper.MetadataUpdateHandler` which would be used in the attribute and it would simply forward the call to the main class. That should not produce any warnings at all. Personally I would probably prefer that solution - it doesn't feel that bad.

I think the delegate related warnings are for the same reason - linker has some implicit rules about marking things on delegate types, so applying `All` to a type which declares a delegate would likely produce these warnings. If the attribute didn't have `All` this would probably go away (but the solution with a static subclass would also fix this).

Reply:
The tow IL2026 warnings are also very likely due to the `MetadataUpdateHandler` attribute, this time on `RouteValueDictionary`. Again, using a static subclass as the handler would solve the problem - but I would question the annotation on the attribute - it clearly should not need constructors or nested types.

Reply:
I created https://github.com/dotnet/linker/issues/2668 to provide a better location for these warnings which should make them easier to diagnose.